File: core/src/main/java/ru/surfstudio/android/core/app/connection/ConnectionProvider.java
package ru.surfstudio.android.core.app.connection;

import android.content.Context;
import android.content.IntentFilter;

import io.reactivex.Observable;

import static android.net.ConnectivityManager.CONNECTIVITY_ACTION;

/**
 * Provider, позволяющий подписаться на событие изменения состояния соединения
 */
public class ConnectionProvider {

    private ConnectionReceiver receiver;

    public ConnectionProvider(Context context) {
        this.receiver = new ConnectionReceiver(context);

        IntentFilter intentFilter = new IntentFilter();
        intentFilter.addAction(CONNECTIVITY_ACTION);
        context.registerReceiver(receiver, intentFilter);
    }

    public Observable<Boolean> observeConnectionChanges() {
        return receiver.observeConnectionChanges();
    }

    public boolean isConnected() {
        return receiver.isConnected();
    }

    public boolean isDisconnected() {
        return !receiver.isConnected();
    }
}

File: connection/lib-connection/src/main/java/ru/surfstudio/android/connection/ConnectionProvider.kt
package ru.surfstudio.android.connection

import android.content.Context
import android.net.NetworkInfo
import io.reactivex.Observable

/**
 * Provider, позволяющий подписаться на событие изменения состояния соединения
 */
interface ConnectionProvider {

    fun observeConnectionChanges(): Observable<Boolean>

    fun isConnected(): Boolean

    fun isDisconnected(): Boolean

    fun isConnectedFast(): Boolean

    /**
     * Проверка на подключение к Wi-Fi
     *
     * @return подключен ли девайс к Wi-Fi, или к мобильной сети
     */
    fun isConnectedToWifi(): Boolean

    fun getNetworkInfo(context: Context): NetworkInfo?
}